File: 0001-TwoSum/Solution.go
/*  Solution: Hash Map
    Time complexity:    O(n)
    Space complexity:   O(n)
 */

func twoSum(nums []int, target int) []int {
  numsChecked := make(map[int]int)
  for index, value := range nums {
    complement := target - value
    if compIndex, ok := numsChecked[complement]; ok {
      return []int{compIndex, index}
    } else {
      numsChecked[value] = index
    }
  }
  return []int{-1,-1}
}

File: 0001-TwoSum/Solution.py
# Solution: Dictionary (Hash Map)
# Time Complexity:  O(n)
# Space Complexity: O(n)

class Solution:
  def twoSum(self, nums: List[int], target: int) -> List[int]:
    numsTracked = {}
    for index, value in enumerate(nums):
      complement = target - value
      if complement in numsTracked:
        return [numsTracked[complement], index]
      else:
        numsTracked[value] = index
